sphinx_needs/directives/need.py

```diff
@@ -376,29 +380,23 @@ def process_need_nodes(app: Sphinx, doctree: nodes.document, fromdocname: str) -
         return
 
     env = app.env
+    needs_data = SphinxNeedsData(env)
 
     # If no needs were defined, we do not need to do anything
-    if not hasattr(env, "needs_all_needs"):
+    if not needs_data.get_or_create_needs():
         return
 
-    # Call dynamic functions and replace related node data with their return values
-    resolve_dynamic_values(env)
-
-    # Apply variant handling on options and replace its values with their return values
-    resolve_variants_options(env)
-
-    # check if we have dead links
-    check_links(env)
-
-    # Create back links of common links and extra links
-    for links in needs_config.extra_links:
-        create_back_links(env, links["option"])
-
-    process_constraints(app)
+    if not needs_data.needs_is_post_processed:
+        resolve_dynamic_values(env)
+        resolve_variants_options(env)
+        check_links(env)
+        create_back_links(env)
+        process_constraints(app)
+        extend_needs_data(app)
+        needs_data.needs_is_post_processed = True
 
-    # We call process_needextend here by our own, so that we are able
-    # to give print_need_nodes the already found need_nodes.
-    process_needextend(app, doctree, fromdocname)
+    for extend_node in doctree.findall(Needextend):
```

Comment:
This is split out from the original `process_needextend` function, since it is not part of the needs data post-processing